src/libcollectionstest/cow_str.rs

```diff
@@ -0,0 +1,65 @@
+// Copyright 2012-2013-2014 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::borrow::Cow;
+
+// check that Cow<'a, str> implements addition
+#[test]
+fn check_cow_add() {
+    borrowed1 = Cow::Borrowed("Hello, ");
+    borrowed2 = Cow::Borrowed("World!");
+    borrow_empty = Cow::Borrowed("");
+
+    owned1 = Cow::Owned("Hi, ".into());
+    owned2 = Cow::Owned("Rustaceans!".into());
+    owned_empty = Cow::Owned("".into());
+
+    assert_eq!("Hello, World!", borrowed1 + borrowed2);
+    assert_eq!("Hello, Rustaceans!", borrowed1 + owned2);
+
+    assert_eq!("Hello, World!", owned1 + borrowed2);
+    assert_eq!("Hello, Rustaceans!", owned1 + owned2);
```

Comment:
These two should presumably start with "Hi, ".

Also `owned1` and `owned2` have been consumed by this point so will need cloning or something earlier.